src/lexparse/lexer.go

```diff
@@ -3,73 +3,228 @@ package lexparse
 import (
   "fmt"
   "strings"
-)
-
-type tokenType int
-
-const (
-  SymbolT tokenType = iota
-
-  StringT tokenType = iota
-  CharT tokenType = iota
-  IntT tokenType = iota
-  FloatT tokenType = iota
-  BoolT tokenType = iota
-  NullT tokenType = iota
-
-  BinopT tokenType = iota // + - * / ^ % == != <= >=
-  SetterT tokenType = iota // = += -= *= /= ^= %=
-  PrefixT tokenType = iota // ! ~
-  Suffix tokenType = iota // $ #
-
-  LParT tokenType = iota
-  RParT tokenType = iota
-  LSquareBrT tokenType = iota
-  RSquareBrT tokenType = iota
-  LCurlyBrT tokenType = iota
-  RCurlyBrT tokenType = iota
-  LAngleT tokenType = iota
-  RAngleT tokenType = iota
-
-  DotT tokenType = iota
-  QMarkT tokenType = iota
-  EllipsisT tokenType = iota
-  CommaT tokenType = iota
-  SemicolonT tokenType = iota
-  ColonT tokenType = iota
+  "os"
+  "io/ioutil"
+  "unicode"
 )
 
 type Line struct {
   Num int
-  Value string
+  Value []rune
   Filename string
 }
 
 type Token struct {
-  Line Line
+  Line *Line
   Col int
   TType tokenType
-  Value string
+  Value []rune
 }
 
-func lexerPanic(token Token, message string) {
-  s := fmt.Sprintf(
-    "Teko Parser Error (%s:%d)\n%s\n%s%s\n%s",
-    token.Line.Filename,
-    token.Col,
-    token.Line.Value,
-    strings.Repeat(" ",token.Col),
-    strings.Repeat("^",len(token.Value)),
+func lexparsePanic(line *Line, col int, width int, message string) {
+  fmt.Printf(
+    "Teko Parser Error (%s:%d)\n%s\n%s%s\n%s\n",
+    line.Filename,
+    col,
+    string(line.Value),
+    strings.Repeat(" ",col),
+    strings.Repeat("^",width),
     message,
   )
-  panic(s)
+  os.Exit(0)
+}
+
+func tokenPanic(token Token, message string) {
+  lexparsePanic(token.Line, token.Col, len(token.Value), message)
+}
+
+type Lexer struct {
+  Line *Line
+  Col int
+  CurrentBlob []rune
+  InBlockComment bool
+  CurrentTType tokenType
+}
+
+func LexFile(filename string) []Token {
```

Comment:
I think most production-grade lexers return one token at a time to avoid the cost of holding the entire file's tokens in memory. The parser calls `lexer.nextToken()` whenever it needs a new token. In other words, the lexer and parser work in tandem instead of in sequence.

I imagine you'd rarely notice the performance hit. Just something to think about.

Reply:
I see - good thought.